src/token.rs

```diff
@@ -0,0 +1,221 @@
+/// The model representing a row in the `api_tokens` database table.
+#[derive(Clone, Debug, PartialEq, Eq, Identifiable, Queryable)]
+pub struct ApiToken {
+    pub id: i32,
+    pub user_id: i32,
+    pub token: String,
+    pub name: String,
+    pub created_at: Timespec,
+    pub last_used_at: Option<Timespec>,
+}
+
+/// The serialization format for the `ApiToken` model.
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct EncodableApiToken {
+    pub id: i32,
+    pub name: String,
+    pub token: Option<String>,
+    pub created_at: String,
+    pub last_used_at: Option<String>,
+}
+
+impl ApiToken {
+    /// Generates a new named API token for a user
+    pub fn insert(conn: &PgConnection, user_id: i32, name: &str) -> CargoResult<ApiToken> {
+        #[table_name = "api_tokens"]
+        #[derive(Insertable, AsChangeset)]
+        struct NewApiToken<'a> {
+            user_id: i32,
+            name: &'a str,
+        }
+
+        diesel::insert(&NewApiToken {
+            user_id: user_id,
+            name: name,
+        }).into(api_tokens::table)
+            .get_result::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    /// Deletes the provided API token if it belongs to the provided user
+    pub fn delete(conn: &PgConnection, user_id: i32, id: i32) -> CargoResult<()> {
+        diesel::delete(api_tokens::table.find(id).filter(
+            api_tokens::user_id.eq(user_id),
+        )).execute(conn)?;
+        Ok(())
+    }
+
+    pub fn find_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<Vec<ApiToken>> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .order(api_tokens::created_at.desc())
+            .load::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    pub fn count_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<u64> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .count()
+            .get_result::<i64>(conn)
+            .map(|count| count as u64)
+            .map_err(From::from)
+    }
+
+    /// Converts this `ApiToken` model into an `EncodableApiToken` for JSON
+    /// serialization.
+    pub fn encodable(self) -> EncodableApiToken {
+        EncodableApiToken {
+            id: self.id,
+            name: self.name,
+            token: None,
+            created_at: ::encode_time(self.created_at),
+            last_used_at: self.last_used_at.map(::encode_time),
+        }
+    }
+
+    /// Converts this `ApiToken` model into an `EncodableApiToken` including
+    /// the actual token value for JSON serialization.
+    pub fn encodable_with_token(self) -> EncodableApiToken {
+        EncodableApiToken {
+            id: self.id,
+            name: self.name,
+            token: Some(self.token),
+            created_at: ::encode_time(self.created_at),
+            last_used_at: self.last_used_at.map(::encode_time),
+        }
+    }
+}
+
+struct BadRequest<T: CargoError>(T);
```

Comment:
Also -- reading further down, this is only ever passed `human`. Should it just take a `String` instead until there's evidence that it needs to take something else?